### PR TITLE
Stwobackend prover skeleton 

### PR DESCRIPTION
**Pull request by ShuangWu121**
### **Progress on Integrating `Stwo`**

This PR makes incremental progress towards integrating the `stwo` prover, refer to issue #1833 . It includes the following changes:

- **Add `stwo` feature to `Cargo.toml`:**  
  The `Cargo.toml` files for the `cli`, `pipeline`, and `backend` components have been updated to include the `stwo` feature.

- **Implement `StwoProverFactory`:**  
  A skeleton structure for `StwoProverFactory` has been introduced. This structure implements the `BackendFactory` trait.

- **Implement `StwoProver`:**  
  Similarly, a skeleton implementation of `StwoProver` has been created, implementing the `Backend` trait. .

### **Commands to run:**

```bash


# Run the project with the `stwo` feature
cargo run --features stwo pil test_data/pil/fibonacci.pil -o output -f --field m31 --prove-with stwo

### CODE REVIEW

File: backend/src/stwo/mod.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        if prev_proof.is_some() {
+            return Err(Error::NoAggregationAvailable);
+        }
+        let proof = vec![0u8; 10];
+        panic!("prove function is not implement yet");
+        Ok(proof)
+    }
+    #[allow(unused_variables)]
+    fn export_verification_key(&self, output: &mut dyn io::Write) -> Result<(), Error> {
+        Ok(())
```
[the remark this answers is not in the document]
Comment (ShuangWu121) [reply backend/src/stwo/mod.rs]: I like this neat trick, appreciate the tip! now it is ready for re-review

File: backend/Cargo.toml [hunk 2]
```diff
@@ -36,6 +37,9 @@ halo2_curves = { version = "0.6.1", package = "halo2curves", optional = true}
 snark-verifier = { git = "https://github.com/powdr-labs/snark-verifier", branch = "powdr-halo2", optional = true }
 halo2_solidity_verifier = { git = "https://github.com/powdr-labs/halo2-solidity-verifier", branch = "main", features = ["evm"], optional = true }
 
+#change this to main branch when updatetoolchain branch is merged,the main branch is using "nightly-2024-01-04", not compatiable with plonky3
```
Comment (georgwiese) [inline backend/Cargo.toml:40]:
```suggestion
# TODO: Change this to main branch when the `andrew/dev/update-toolchain` branch is merged,the main branch is using "nightly-2024-01-04", not compatiable with plonky3
```
Do you mean [this branch](https://github.com/starkware-libs/stwo/compare/dev...andrew/dev/update-toolchain)? I didn't find an `updatetoolchain` branch.

File: backend/src/stwo/mod.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+        options: BackendOptions,
+    ) -> Result<Box<dyn crate::Backend<F>>, Error> {
+        if pil.degrees().len() > 1 {
+            return Err(Error::NoVariableDegreeAvailable);
+        }
+        let fixed = Arc::new(
+            get_uniquely_sized_cloned(&fixed).map_err(|_| Error::NoVariableDegreeAvailable)?,
+        );
+        let stwo = Box::new(StwoProver::new(pil, fixed, setup)?);
+        info!("StwoProverFactory is not complete yet");
```
Comment (georgwiese) [inline backend/src/stwo/mod.rs:38]:
```suggestion
```

I think the "not implemented" panic should be enough :)

File: backend/src/stwo/mod.rs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        Ok(stwo)
+    }
+}
+
+impl<T: FieldElement> Backend<T> for StwoProver<T> {
+    #[allow(unused_variables)]
+    fn verify(&self, proof: &[u8], instances: &[Vec<T>]) -> Result<(), Error> {
+        assert!(instances.len() == 1);
+        Ok(())
```
Comment (georgwiese) [inline backend/src/stwo/mod.rs:47]:
```suggestion
        unimplemented!()
```
No?

**Review by georgwiese — APPROVED** [review]
LGTM!

**Pull request merged**